### PR TITLE
Determine whether repeated tasks will be run.

### PR DESCRIPTION
**Pull request by magiWei**
When user run a task from python client, API Server will check whether there is a task which run successfully and contains the same simulation file by file md5. if exist repeated task, API server will  delete the new task and s3 file and throw specified error and python client needs to deal the error.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+- Determine whether repeated tasks will be run.
```
Comment (tylerflex) [inline CHANGELOG.md:9]: I think this should be more specific, perhaps
```
- If a simulation is detected as being run previously, downloads the results from the original simulation instead.
```

File: tidy3d/web/core/exceptions.py [hunk 2]
```diff
@@ -10,3 +10,13 @@ def __init__(self, message: str = None):
...
         log.error(message)
+
+
+class TaskRepeatedError(Exception):
+    """Repeated task in tidy3d"""
+
+    def __init__(self, message: str = None):
+        """Log just the error message and then raise the Exception."""
+        # log = get_logger()
+        super().__init__(message)
```
Comment (tylerflex) [inline tidy3d/web/core/exceptions.py:21]: I dont think this line of code does anything different from the original init. I think we can delete this whole method.

File: tidy3d/web/core/http_util.py [hunk 3]
```diff
@@ -129,6 +131,10 @@ def wrapper(*args, **kwargs):
         if resp.status_code != ResponseCodes.OK.value:
             if resp.status_code == ResponseCodes.NOT_FOUND.value:
                 return None
+            if resp.status_code == ResponseCodes.UNPROCESSABLE.value and resp.text:
+                result = resp.json()
+                if str(ResponseCodes.TASK_REPEATED.value) == result["code"]:
```
Comment (tylerflex) [inline tidy3d/web/core/http_util.py:136]: what's the reason for `str()` casting here and not in the other `if` statements above?

Comment (tylerflex) [inline tidy3d/web/core/http_util.py:136]: I'm worried about obscure errors happening if `"code"` or `"detail"` are not found in `result`. Is there a way to handle this more gracefully?  maybe
```py
if any(key not in result for key in ("code", "detail")):
    # do task repeated logic
else:
    # not sure, either raise a new WebError about missing keys or just continue?
```

```

File: tidy3d/web/api/webapi.py [hunk 4]
```diff
@@ -88,15 +89,35 @@ def run(
...
+            worker_group=worker_group,
+        )
+        monitor(task_id, verbose=verbose)
+        return load(
+            task_id=task_id,
+            path=path,
+            verbose=verbose,
+            progress_callback=progress_callback_download,
+        )
+    except TaskRepeatedError as e:
```
Comment (tylerflex) [inline tidy3d/web/api/webapi.py:105]: when does this get raised? in `start`? if so, maybe we should put this `except` handling there?  otherwise it will only have this feature if user uses `web.run()`,  not if they do `web.start()` nor if they use `Job.run()`, `Batch.run()`. I think it would be better if we could catch it at all times?

Maybe the issue is that we need a new task_id to continue the loading? if that's the case, maybe a cleaner approach is to allow `web.start` to return an optional `task_id` object and then slightly refactor the webapi internals to use this returned task_id in the subsequent code?

Basically in it's current stage, I think it needs to be refactored to future-proof and make this generally work every time `web.start()` is called.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Thanks @magiWei for the PR.

I think there's a bit of changes required, we should also probably get @lei-flex to look at it. My feeling is that we can refactor this handling a bit better inside of the `start()` function for more generality + some other minor comments, mainly about making sure the code is robust if this feature fails for some reason.

My more general concern is with the design of the caching. So in the future, there will never be duplicate tasks stored on our side? Basically every time a new task is made that is a duplicate of another task, it is immediately deleted? This might be a big change to a user's workflow if they are doing something like several parameter sweeps (like in different folders) with duplicated data.

A few discussion points / suggestions:
1. Is there some way to easily make this an optional feature? 
2. An architecture that would make sense to me is to separate the simulation from the data. So we could have duplicate tasks with the same simulation, but only the data for unique simulations would be stored. So if a user submits task1 and task2 with the same simulation. We keep both tasks in the UI but just do a mapping from simulation -> data and grab the common data. To me this seems like an architecture that should be considered, do you have thoughts on this design?

**Comment by magiWei** [comment]
> Thanks @magiWei for the PR.
> 
> I think there's a bit of changes required, we should also probably get @lei-flex to look at it. My feeling is that we can refactor this handling a bit better inside of the `start()` function for more generality + some other minor comments, mainly about making sure the code is robust if this feature fails for some reason.
> 
> My more general concern is with the design of the caching. So in the future, there will never be duplicate tasks stored on our side? Basically every time a new task is made that is a duplicate of another task, it is immediately deleted? This might be a big change to a user's workflow if they are doing something like several parameter sweeps (like in different folders) with duplicated data.
> 
> A few discussion points / suggestions:
> 
> 1. Is there some way to easily make this an optional feature?
> 2. An architecture that would make sense to me is to separate the simulation from the data. So we could have duplicate tasks with the same simulation, but only the data for unique simulations would be stored. So if a user submits task1 and task2 with the same simulation. We keep both tasks in the UI but just do a mapping from simulation -> data and grab the common data. To me this seems like an architecture that should be considered, do you have thoughts on this design?

About #2, We have considered keeping task1 and task2 in DB, and task1 and task2 link to the same s3 file. If so, all operations involving tasks need to determine whether the task links the same s3 file, including: query, delete, modify, copy functions, etc.  the complexity of the API Server system will be greatly increased. So API server throws an error when checking one repeated task to python client.
